[yikjin] iP

yikjin opened the pull request:
Creating this pull request as instructed by the [Week 4 project page](https://nus-cs2113-ay2526s1.github.io/website/schedule/week4/project.html#1-create-a-pr-to-the-upstream-repo-mon-sep-1st-2359).

---

_Note:_ This project is part of [NUS' "CS2113: Software Engineering & Object-Oriented Programming" module](https://nusmods.com/courses/CS2113/software-engineering-object-oriented-programming).

yikjin committed:
chore(setup): add project format and lint config

Add IntelliJ IDEA project-wide code formatting and linting configuration
files in `.idea/`, aligned to the [CS2113 Java coding standard](
https://se-education.org/guides/conventions/java/basic.html) based on
the [Google Java style guide](
https://google.github.io/styleguide/javaguide.html).

Add `.editorconfig` code formatting configuration file to allow other
IDEs to follow the required code style guidelines as well.

yikjin committed:
chore(setup): re-format existing codebase

Reformat existing codebase in IntelliJ IDEA automatically using the code
formatting configuration files.

yikjin committed:
chore(setup): add iP info

Update `README.md` to include my own iP information.

Add `LICENSE.md` to specify that my iP is not an open source project (to
prevent others from copying my iP code to use in their own).

Add `CODEOWNERS` file to specify that I own my iP files.

Add `SECURITY.md` and `CODE_OF_CONDUCT.md` files, lightly modified from
the existing GitHub repository templates.

yikjin committed:
chore(setup): update Git project config

Update `.gitignore` to include more files to be automatically ignored by
Git, without having to manually exclude them.

Add `.gitattributes` to specify correct file types and line endings to
Git.

Add `.gitconfig` to standardise Git project-wide settings for everyone.

yikjin committed:
chore(setup): merge branch 'setup'

Add project code formatting and linting configuration files.

Update existing codebase formatting, and project configuration and
documentation files.

yikjin committed:
feat(Level-0): rename chatbot from Duke to Kuro

Rename all mentions of the chatbot to Kuro from the placeholder name
Duke, except for the section of README preserved from upstream.

yikjin committed:
feat(Level-0): add skeletal greet and quit methods

Add basic greet and quit methods that just prints to the console the
respective messages.

Update the UI test case to new output, so that the test case passes.

yikjin committed:
style: remove `FinalClass` inspection

Disable `FinalClass` inspection ("Class is closed to inheritance"),
since it can conflict with the `NonFinalUtilityClass` inspection
("Utility class is not 'final'").

yikjin committed:
feat(Level-0): merge branch 'branch-Level-0'

Update chatbot name to Kuro from the placeholder name Duke.

Add initial skeletal version that greets the user and exits.

yikjin committed:
feat(Level-1): add command echo to Kuro

Add ability for Kuro to echo commands passed to it after the greeting is
displayed. The `bye` command still prints the exit message without
changes.

yikjin committed:
style: remove `SingleClassImport` inspection

Disable `SingleClassImport` inspection ("Single class import"), since it
warns about imports unnecessarily.

Rename the inspection profile to `Default.xml` from `Project.xml`, since
that is the default name for project-wide inspection profiles.

yikjin committed:
style: fix comments generated at start of line

Update IntelliJ IDEA to generate comments from the indentation level
instead of at the start of the line.

yikjin committed:
chore: remove `.gitignore` in `.idea` directory

Remove unnecessary `.gitignore` file in the `.idea` directory, since all
the rules are already covered in the main `.gitignore` file.

yikjin committed:
feat(Level-1): merge branch 'branch-Level-1'

Add ability for Kuro to echo commands typed in after greeting the user,
while keeping the `bye` command working as before.

yikjin committed:
feat(Level-2): add and list tasks for Kuro

Add ability to add and read tasks for Kuro. Tasks are not saved after
program termination.

yikjin committed:
test: update test to just exiting Kuro

Update the test input and expected output to quit on running Kuro.

yikjin committed:
style: update code styles and inspection profiles

Update the IntelliJ IDEA code styles and inspection profiles align with
the code convention used.

yikjin committed:
feat(Level-2): merge branch 'branch-Level-2'

Add ability to add and read tasks for Kuro. Tasks are not saved after
program termination.

yikjin committed:
feat(Level-3): add mark and unmark tasks as done

Add mark and unmark tasks as done to Kuro. Marking multiple tasks in one
go is supported.

yikjin committed:
style: update code styles and inspection profiles

Update the IntelliJ IDEA code styles and inspection profiles align with
the code convention used.

yikjin committed:
feat(Level-3): merge branch 'branch-Level-3'

Add mark and unmark tasks as done to Kuro. Marking multiple tasks in one
go is supported.

andrewcai8 commented on src/main/java/Kuro.java:
```diff
@@ -0,0 +1,169 @@
+        System.out.println("Bye. Hope to see you again soon!");
+    }
+
+    public static void main(final String[] args) {
+        Kuro.greet();
+        Kuro.startChatSession();
+        Kuro.quit();
+    }
```
I like how the main() calls 3 methods, it keeps the code organized

andrewcai8 commented on src/main/java/Kuro.java:
```diff
@@ -0,0 +1,169 @@
+        }
+    }
+
+    private static void quit() {
+        System.out.println("Bye. Hope to see you again soon!");
+    }
+
```
you can put quit() method into the "bye" case, i think it makes a bit more sense

andrewcai8 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,35 @@
+    Task(final String name) {
+        this.name = name;
+    }
+
+    char getStatusIcon() {
+        return this.isDone ? 'X' : ' ';
+    }
+
+    public String toString() {
+        return this.name + (this.isDone ? " (done)" : "");
+    }
+
```
the "this." prefix before the variable isn't needed in these methods

andrewcai8 commented on src/main/java/Kuro.java:
```diff
@@ -0,0 +1,169 @@
+
+final class Kuro {
+    private static final int MAX_TASKS = 100;
+    private static final Task[] tasks = new Task[Kuro.MAX_TASKS];
```
No need for the "Kuro." prefix before "Kuro.MAX_TASKS"

chongyaoo commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,35 @@
+    private String name = "";
+    private boolean isDone = false;
+
+    Task() {
```
No need for empty constructor because cannot initialize empty Task

chongyaoo commented on src/main/java/Kuro.java:
```diff
@@ -0,0 +1,169 @@
+        }
+    }
+
+    private static void setTaskDone(final String[] taskIds) {
```
Logic for setTaskDone and setTaskUndone are the same. Consider passing in an additional boolean parameter to determine whether setting or unsetting, then implementing the logic for each case using one setTask(bool set, final String[] taskIds) function.

fookang commented on src/main/java/Kuro.java:
```diff
@@ -0,0 +1,169 @@
+    private static int latestTaskIndex = 0;
+
+    private static void greet() {
+        final String nameLogo = """
```
I like how you separate the logo, it keeps the code more organized

fookang commented on src/main/java/Kuro.java:
```diff
@@ -0,0 +1,169 @@
+                        ? inputTokens[1].split(" ")
+                        : new String[] {};
+
+                switch (inputCommand) {
```
Good job on following the coding standards for switch case

fookang commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,35 @@
+final class Task {
+    private String name = "";
+    private boolean isDone = false;
```
Good job for following naming convention

fookang commented on src/main/java/Kuro.java:
```diff
@@ -0,0 +1,169 @@
+        for (final String taskIdString : taskIds) {
+            final int taskId;
+
+            try {
```
Good job on following the coding standards for try-catch

chongyaoo commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,35 @@
+final class Task {
```
Good to separate the Task class out

yikjin committed:
fix: add tasks with spaces in between

Fix an issue where adding a task with spaces in between (e.g. a
sentence) causes only part of the task (up till the first space) to be
added.

yikjin committed:
refactor: clean up `Kuro.java` and `Task.java`

Refactor the Java source files.

Add additional beginning newline and prompt symbol in chat with Kuro, so
that the input and output can be differentiated easier.

yikjin committed:
style: configure arrangement rules and annotations

Update the IntelliJ IDEA Java code arrangement grouping rules, and make
missing `@Override` annotations show up as warnings.

yikjin committed:
refactor: merge branch 'ip-review-refactor'

Fix adding task with spaces only adding the first word.

Refactor Java source files.

Update IntelliJ IDEA Java code arrangement grouping rules, and always
show missing `@Override` as warnings.

chongyaoo commented on .gitattributes:
```diff
@@ -0,0 +1,150 @@
+# <https://github.com/gitattributes/gitattributes/blob/master/Common.gitattributes>
```
is .gitattributes essential to commit? if not essential, i would suggest not committing this to the online repo, so as to make the file size of the repo smaller. Same goes for any other non-essential files too.

chongyaoo commented on text-ui-test/runtest.bat:
```diff
@@ -15,7 +15,7 @@ IF ERRORLEVEL 1 (
 REM no error here, errorlevel == 0
```
You could write some test cases here in runtest.bat to test the output of your code!

chongyaoo commented on src/main/java/Kuro.java:
```diff
@@ -0,0 +1,138 @@
+        System.out.println("What can I do for you?");
+    }
+
+    private static void startChatSession() {
```
Separation of the cases into each own function makes the code clean and well-structured. Nice work!

chongyaoo reviewed (commented):
Overall great job in the code structure! But there are too many files in the repo, if they are non-essential I would suggest leaving them out of the commit.

yikjin committed:
feat(Level-4): add task types and split to classes

Add todo, deadline, and event task types to Kuro.

Refactor majority of the code to use OOP, splitting into many different
classes to handle the different parts of the program.

yikjin committed:
test(Level-4): update to reflect latest behaviour

Update the integration test to correctly reflect the latest program
features and output.

yikjin committed:
style(Level-4): drop unnecessary inspection rules

Allow use of private fields if in the same class.

Remove suggestion to convert utility classes to `enum`s.

Update `.editorconfig` to match project code styles, disabling usage of
fully qualified class names.

yikjin committed:
feat(Level-4): merge branch 'branch-Level-4'

Add todo, deadline, and event task types to Kuro.

Refactor majority of the code to use OOP, splitting into many different
classes to handle the different parts of the program.

Update the integration test to correctly reflect the latest program
features and output.

Remove unnecessary inspection rules.

yikjin committed:
style: align more closely with coding standard

Remove unnecessary class name and `this` qualifiers.

Rename `Kuro.java` to `Main.java`, making the program entry point more
obvious.

yikjin committed:
feat(Level-5): handle errors using exceptions

Update error handling to make use of custom exceptions, instead of just
printing the error message to the console.

yikjin committed:
feat(Level-5): merge branch 'branch-Level-5'

Update error handling code to use custom exceptions, instead of just
printing the error message to the console.

Update code style to better align with the course coding standard.

yikjin committed:
refactor(A-Packages): move classes into packages

Update all Java classes to be in user-defined packages
(`arpa.home.yikjin.app.kuro.*`).

yikjin committed:
chore: add IntelliJ IDEA run configuration files

Add `Main.class` and `iP.jar` run configuration files for IntelliJ IDEA.

Update `.gitignore` to not ignore some files related to run
configurations. They were previously excluded because the template
`.gitignore` file assumes Gradle or Maven was in use, which is not the
case.

yikjin committed:
refactor(A-Packages): merge branch 'branch-A-Packages'

Update all Java classes to be in user-defined packages
(`arpa.home.yikjin.app.kuro.*`).

Add IntelliJ IDEA run configuration files.

Update `.gitignore` to not exclude some run configuration files.

JoanneHing commented on src/main/java/arpa/home/yikjin/app/kuro/command/InputCommand.java:
```diff
@@ -0,0 +1,27 @@
+package arpa.home.yikjin.app.kuro.command;
+
+enum InputCommand {
+    BYE("bye"), DEADLINE("deadline"), EVENT("event"), LIST("list"), MARK("mark"), TODO("todo"), UNMARK(
+            "unmark"), INVALID("");
+
+    private final String enumValue;
+
+    InputCommand(final String enumValue) {
+        this.enumValue = enumValue;
+    }
```
Good practice on trying out on enums.

JoanneHing commented on src/main/java/arpa/home/yikjin/app/kuro/parser/InputParser.java:
```diff
@@ -0,0 +1,101 @@
+import java.util.regex.Pattern;
+
+final class InputParser {
+    private static final int INIT_POS_ARGS_CAPACITY = 5;
+    private static final int INIT_NAMED_ARGS_CAPACITY = 2;
```
Good job on avoiding using magic numbers.

JoanneHing commented on src/main/java/arpa/home/yikjin/app/kuro/command/Runner.java:
```diff
@@ -0,0 +1,194 @@
+        }
+
+        final String[] posArgs = PARSER.getPositionalArgs();
+        final String posArgsAsString = InputParser.mapArgsToString(posArgs);
+        final Map<String, String[]> namedArgs = PARSER.getNamedArgs();
+
+        switch (command) {
+        case BYE:
+            bye();
+            break;
+        case DEADLINE:
+            addDeadline(posArgsAsString, namedArgs);
+            break;
+        case EVENT:
+            addEvent(posArgsAsString, namedArgs);
+            break;
+        case LIST:
+            listTasks();
+            break;
+        case MARK:
+            markTasksAsDone(posArgs);
+            break;
+        case UNMARK:
+            markTasksAsDone(posArgs, false);
+            break;
+        case TODO:
+            addTodo(posArgsAsString);
+            break;
+        default:
+            throw new InvalidCommandException(rawCommand);
+        }
+    }
```
Good job on using switch-case here instead of if-else blocks

JoanneHing commented on src/main/java/arpa/home/yikjin/app/kuro/exception/input/argument/InvalidTaskIdException.java:
```diff
@@ -0,0 +1,18 @@
+
+public final class InvalidTaskIdException extends AppArgumentException {
+    @Serial
+    private static final long serialVersionUID = 0L;
```
If this is a constant, remember to name the variable in all caps.

JoanneHing commented on src/main/java/arpa/home/yikjin/app/kuro/exception/input/argument/MissingTaskIdsException.java:
```diff
@@ -0,0 +1,14 @@
+
+public final class MissingTaskIdsException extends AppArgumentException {
+    @Serial
+    private static final long serialVersionUID = 0L;
```
Same as above.

JoanneHing commented on src/main/java/arpa/home/yikjin/app/kuro/exception/input/argument/MissingTaskNameException.java:
```diff
@@ -0,0 +1,14 @@
+
+public final class MissingTaskNameException extends AppArgumentException {
+    @Serial
+    private static final long serialVersionUID = 0L;
```
Similar to above.

JoanneHing reviewed (commented):
Overall good progress done so far. Just some constant variable naming missed out on the Java coding naming standards.

yikjin committed:
feat(Level-6): add delete task function

Add support for deleting tasks from the list.

yikjin committed:
feat(Level-6): merge branch 'branch-Level-6'

Add support for deleting tasks from the list.

yikjin committed:
chore: remove empty `package-info.java` files

Remove empty `package-info.java` files, as they do not contain any
documentation. They are there to suppress the warning messages of the
missing `package-info.java` documentation files, which should not be
suppressed if there are no documentation in the files anyway.

yikjin committed:
merge: merge branch 'master' into branch-Level-7

Update `branch-Level-7` branch to latest `master` commits.

yikjin committed:
feat(Level-7): add saving of tasks to file

Add saving and updating to an external file.

yikjin committed:
chore: add '/data' directory to ignore

Add the '/data' directory to both '.gitignore' and 'iP.iml' project
files. The '/data' directory contains the program data files.

yikjin committed:
style: update 'ImplicitNumericConversion' inspection

Update the 'ImplicitNumericConversion' inspection rule to enable both
'ignoreWideningConversions' and 'ignoreCharConversions'.

yikjin committed:
style: add annotation processing

Add annotation processing to IntelliJ IDEA compiler configuration.

yikjin committed:
merge: merge branch 'branch-Level-7'

Add saving and updating of tasks to external data file.

yikjin committed:
refactor(A-MoreOOP): refactor classes and packages

Refactor various classes and packages to be more OOP.

yikjin committed:
merge(A-MoreOOP): merge branch 'A-MoreOOP'

Refactor classes and packages to be more OOP.

yikjin committed:
docs(A-JavaDoc): add JavaDoc comments

Add JavaDoc comments to all classes and methods.

yikjin committed:
docs(A-UserGuide): add Kuro user guide

Add Kuro user guide.

yikjin committed:
merge(A-UserGuide): merge branch 'A-UserGuide'

Create Kuro user guide.

yikjin committed:
build(A-Release): add Gradle build automation

Add Gradle build automation to build project.

yikjin committed:
merge(A-Release): merge branch 'A-Release'

Add Gradle build automation to project.

yikjin committed:
merge: merge pull request #2 from yikjin/A-JavaDoc

docs(A-JavaDoc): add JavaDoc comments
